ssl/statem/statem_clnt.c

```diff
@@ -1128,6 +1129,7 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
             if (!ssl_get_new_session(s, 0)) {
                 goto f_err;
             }
+            s->session->ssl_version = s->version;
```

Comment:
One line down. :-)

The thing you want is for the fresh session to have the same version as the established version. The problem is that, if `session_id_length` is 0---that is, you weren't offering a session at all---we don't call `ssl_get_new_session` here. That means `s->session` is initialized from line 831. But that's before version negotiation!

Thus, the original code "fixed" `ssl_session` afterwards. Because in not all cases was `ssl_session` correct. And this is why I think it's better---but more involved, so it's not the minimal fix---to defer line 831 to here. That way `s->session` has the right version to begin with, rather than the wrong one and then a fixup.

Is that clearer? I feel like I haven't been explaining the issue very well.